rust-runtime/aws-smithy-async/src/future/pagination_stream.rs

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Types to support stream-like operations for paginators.
+
+use crate::future::rendezvous;
+use pin_project_lite::pin_project;
+use std::fmt;
+use std::future::poll_fn;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pub mod collect;
+
+/// A wrapper around [`FnStream`].
+///
+/// This type provides the same set of methods as [`FnStream`], but that is meant to be used
+/// internally and not by external users.
+#[derive(Debug)]
+pub struct PaginationStream<Item>(FnStream<Item>);
+
+impl<Item> PaginationStream<Item> {
+    /// Creates a `PaginationStream` from the given [`FnStream`].
+    pub fn new(stream: FnStream<Item>) -> Self {
+        Self(stream)
+    }
+
+    /// Consumes and returns the next `Item` from this stream.
+    pub async fn next(&mut self) -> Option<Item> {
+        self.0.next().await
+    }
+
+    /// Consumes this stream and gathers elements into a collection.
+    pub async fn collect<T: collect::sealed::Collectable<Item>>(self) -> T {
+        self.0.collect().await
+    }
+}
+
+impl<T, E> PaginationStream<Result<T, E>> {
+    /// Yields the next item in the stream or returns an error if an error is encountered.
+    pub async fn try_next(&mut self) -> Result<Option<T>, E> {
+        self.next().await.transpose()
+    }
+
+    /// Convenience method for `.collect::<Result<Vec<_>, _>()`.
+    pub async fn try_collect(self) -> Result<Vec<T>, E> {
```

Comment:
we _may_ want to make `Vec` a generic here so that you collect other types in the future. I don't feel storngly though.

Reply:
A bit torn on that, but I'd leave `try_collect` generic free for now. If someone wants a different collection type while using `try_collect`, they can use `collect` with a turbo fish instead.